File: bindings/python/rings.cpp
#include <boost/python.hpp>

#include "../../src/chemist/molecule.h"
#include "../../src/algorithms/cycles.h"
#include "../../src/ring.h"
#include "common.h"

using Helium::Chemist::Atom;
using Helium::Chemist::Bond;
using Helium::Chemist::Molecule;
using namespace boost::python;

Helium::Size cyclomatic_number_1(const Molecule &mol, Helium::Size numComponents)
{
  return Helium::cyclomatic_number(mol, numComponents);
}

Helium::Size cyclomatic_number_2(const Molecule &mol)
{
  return Helium::cyclomatic_number(mol);
}

object cycle_membership(const Molecule &mol)
{
  std::vector<bool> cycleAtoms;
  std::vector<bool> cycleBonds;

  Helium::cycle_membership(mol, cycleAtoms, cycleBonds);

  return make_tuple(cycleAtoms, cycleBonds);
}

Helium::RingSet<Molecule> relevant_cycles_1(const Molecule &mol, Helium::Size cyclomaticNumber,
    const list &cyclicAtoms, const list &cyclicBonds)
{
  return Helium::relevant_cycles_vismara(mol, cyclomaticNumber,
      vector_from_list<bool>(cyclicAtoms), vector_from_list<bool>(cyclicBonds));
}

Helium::RingSet<Molecule> relevant_cycles_2(const Molecule &mol)
{
  return Helium::relevant_cycles_vismara(mol);
}

Atom Ring_atom(const Helium::Ring<Molecule> &ring, std::size_t index)
{
  if (index >= ring.size())
    throw std::runtime_error("Invalid ring atom index");
  return ring.atom(index);
}

Bond Ring_bond(const Helium::Ring<Molecule> &ring, std::size_t index)
{
  if (index >= ring.size())
    throw std::runtime_error("Invalid ring bond index");
  return ring.bond(index);
}

const Helium::Ring<Molecule>& RingSet_ring(const Helium::RingSet<Molecule> &rings, std::size_t index)
{
  if (index >= rings.size())
    throw std::runtime_error("Invalid ring index");
  return rings.ring(index);
}

void export_rings()
{

  def("cyclomatic_number", &cyclomatic_number_1);
  def("cyclomatic_number", &cyclomatic_number_2);
  def("cycle_membership", &cycle_membership);
  def("relevant_cycles", &relevant_cycles_1);
  def("relevant_cycles", &relevant_cycles_2);

  class_<Helium::Ring<Molecule> >("Ring", no_init)
    .def("size", &Helium::Ring<Molecule>::size)
    .def("__len__", &Helium::Ring<Molecule>::size)
    .def("atoms", &Helium::Ring<Molecule>::atoms, return_value_policy<copy_const_reference>())
    .def("atom", &Ring_atom)
    .def("bonds", &Helium::Ring<Molecule>::bonds, return_value_policy<copy_const_reference>())
    .def("bond", &Ring_bond)
    .def("containsAtom", &Helium::Ring<Molecule>::containsAtom)
    .def("containsBond", &Helium::Ring<Molecule>::containsBond)
    ;

  class_<Helium::RingSet<Molecule> >("RingSet", init<const Molecule&>())
    .def("size", &Helium::RingSet<Molecule>::size)
    .def("__len__", &Helium::RingSet<Molecule>::size)
    .def("rings", &Helium::RingSet<Molecule>::rings, return_value_policy<copy_const_reference>())
    .def("ring", &RingSet_ring, return_internal_reference<>())
    .def("isAtomInRing", &Helium::RingSet<Molecule>::isAtomInRing)
    .def("isBondInRing", &Helium::RingSet<Molecule>::isBondInRing)
    .def("isAtomInRingSize", &Helium::RingSet<Molecule>::isAtomInRingSize)
    .def("isBondInRingSize", &Helium::RingSet<Molecule>::isBondInRingSize)
    .def("numRingNbrs", &Helium::RingSet<Molecule>::numRingNbrs)
    .def("numRings", &Helium::RingSet<Molecule>::numRings)
    ;
}

File: test/python/rings.py
import helium
import unittest

SMILES = helium.Smiles()

class TestRings(unittest.TestCase):

    def test_cyclomatic_number(self):
        mol = helium.Molecule()
        SMILES.read('CC1CCCCC1C', mol)

        self.assertEqual(1, helium.cyclomatic_number(mol, 1))
        self.assertEqual(1, helium.cyclomatic_number(mol))

    def test_cycle_membership(self):
        mol = helium.Molecule()
        SMILES.read('CC1CC1C', mol)

        atoms, bonds = helium.cycle_membership(mol)

        self.assertListEqual([False, True, True, True, False], atoms)
        self.assertListEqual([False, True, True, True, False], bonds)

    def test_relevant_cycles(self):
        mol = helium.Molecule()
        SMILES.read('c1ccccc1-c1ccccc1', mol)

        rings = helium.relevant_cycles(mol)
        self.assertEqual(2, len(rings))

        atoms, bonds = helium.cycle_membership(mol)
        rings = helium.relevant_cycles(mol, 1, atoms, bonds)
        self.assertEqual(2, len(rings))

    def test_ringset(self):
        mol = helium.Molecule()
        SMILES.read('CCC1CC1', mol)
        rings = helium.relevant_cycles(mol)
        self.assertTrue(isinstance(rings, helium.RingSet))

        self.assertEqual(1, rings.size())
        self.assertEqual(1, len(rings))

        self.assertEqual(1, len(rings.rings()))
        self.assertTrue(isinstance(rings.ring(0), helium.Ring))

        self.assertFalse(rings.isAtomInRing(mol.atom(0)))
        self.assertTrue(rings.isAtomInRing(mol.atom(2)))

        self.assertFalse(rings.isBondInRing(mol.bond(0)))
        self.assertTrue(rings.isBondInRing(mol.bond(2)))

        self.assertFalse(rings.isAtomInRingSize(mol.atom(2), 4))
        self.assertTrue(rings.isAtomInRingSize(mol.atom(2), 3))

        self.assertFalse(rings.isBondInRingSize(mol.bond(2), 4))
        self.assertTrue(rings.isBondInRingSize(mol.bond(2), 3))

        self.assertEqual(2, rings.numRingBonds(mol.atom(2)))
        self.assertEqual(2, rings.numRingNbrs(mol.atom(2)))
        self.assertEqual(1, rings.numRings(mol.atom(2)))

    def test_ring(self):
        mol = helium.Molecule()
        SMILES.read('CCC1CC1', mol)
        rings = helium.relevant_cycles(mol)
        ring = rings.ring(0)

        self.assertEqual(3, ring.size())
        self.assertEqual(3, len(ring))
        self.assertEqual(3, len(ring.atoms()))
        self.assertEqual(3, len(ring.bonds()))

        self.assertEqual(4, ring.atom(0).index())
        self.assertEqual(3, ring.atom(1).index())
        self.assertEqual(2, ring.atom(2).index())

        self.assertEqual(3, ring.bond(0).index())
        self.assertEqual(2, ring.bond(1).index())
        self.assertEqual(4, ring.bond(2).index())

        self.assertFalse(ring.containsAtom(mol.atom(0)))
        self.assertTrue(ring.containsAtom(mol.atom(2)))

        self.assertFalse(ring.containsBond(mol.bond(0)))
        self.assertTrue(ring.containsBond(mol.bond(2)))

if __name__ == '__main__':
    unittest.main()
